### PR TITLE
fix: force user to relogin on session expiration closes #2342

### PR DESCRIPTION
**Pull request by gen1us2k**
## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
      and signed the CLA.
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added necessary documentation within the code base (if
      appropriate).

### CODE REVIEW

**Review by aeneasr — CHANGES_REQUESTED** [review]
Could you please add a test for this? :)

File: cloudx/handler.go [hunk 1]
```diff
@@ -232,6 +232,27 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
 
 	if len(c.SessionToken) > 0 {
 		_, _ = fmt.Fprintf(h.verboseErrWriter, "You are authenticated as: %s\n", c.IdentityTraits.Email)
+		client, err := newKratosClient()
+		if err != nil {
+			return nil, err
+		}
+		_, resp, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if resp.StatusCode == http.StatusUnauthorized || err != nil {
+			ok, err := cmdx.AskScannerForConfirmation(fmt.Sprintf("Your CLI session has expired. Do you wish to log in again with \"%s\"?", c.IdentityTraits.Email), h.stdin, h.verboseErrWriter)
```
Comment (zepatrik) [inline cloudx/handler.go:241]:
```suggestion
			ok, err := cmdx.AskScannerForConfirmation(fmt.Sprintf("Your CLI session has expired. Do you wish to log in again as \"%s\"?", c.IdentityTraits.Email), h.stdin, h.verboseErrWriter)
```

File: cloudx/handler.go [hunk 2]
```diff
@@ -232,6 +232,27 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
 
 	if len(c.SessionToken) > 0 {
 		_, _ = fmt.Fprintf(h.verboseErrWriter, "You are authenticated as: %s\n", c.IdentityTraits.Email)
```
Comment (zepatrik) [inline cloudx/handler.go:234]: Potentially move this to after the check? Would make more sense form a UX perspective.

File: cloudx/handler.go [hunk 3]
```diff
@@ -232,6 +232,27 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
+			return nil, err
+		}
+		_, resp, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if resp.StatusCode == http.StatusUnauthorized || err != nil {
+			ok, err := cmdx.AskScannerForConfirmation(fmt.Sprintf("Your CLI session has expired. Do you wish to log in again with \"%s\"?", c.IdentityTraits.Email), h.stdin, h.verboseErrWriter)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				if err := h.SignOut(); err != nil {
```
Comment (zepatrik) [inline cloudx/handler.go:246]: Does this only delete the current session but not the email address?

File: cloudx/handler.go [hunk 4]
```diff
@@ -232,6 +232,27 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
+					return nil, err
+				}
+				c, err := h.Authenticate()
```
Comment (zepatrik) [inline cloudx/handler.go:249]: And will then here the email address be pre-filled?

**Review by zepatrik — CHANGES_REQUESTED** [review]

File: cloudx/handler.go (hunk 3 above)
Comment (gen1us2k) [reply cloudx/handler.go:246]: It removes everything and creates `AuthContext` with null values

**Comment by aeneasr** [comment]
Tests are currently failing - please take a look :)

**Comment by gen1us2k** [comment]
@aeneasr Fixed tests.

File: cloudx/handler.go [hunk 5]
```diff
@@ -480,14 +500,14 @@ func (h *SnakeCharmer) ListProjects() ([]cloud.Project, error) {
 	if err != nil {
 		return nil, err
 	}
-
```
Comment (aeneasr) [inline cloudx/handler.go:483]: Can you please revert these newline changes? They make the code less readable. We try to separate blocks of logic using a singular newline :)

File: cloudx/handler.go [hunk 6]
```diff
@@ -231,6 +231,27 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
 	}
 
 	if len(c.SessionToken) > 0 {
+		client, err := newKratosClient()
+		if err != nil {
+			return nil, err
+		}
+		_, resp, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if resp.StatusCode == http.StatusUnauthorized || err != nil {
+			ok, err := cmdx.AskScannerForConfirmation(fmt.Sprintf("Your CLI session has expired. Do you wish to log in again as \"%s\"?", c.IdentityTraits.Email), h.stdin, h.verboseErrWriter)
```
Comment (aeneasr) [inline cloudx/handler.go:240]: If the quiet flag is set, this should error and not prompt!

File: cloudx/handler.go (hunk 4 above)
Comment (aeneasr) [reply cloudx/handler.go:248]: yes

**Review by aeneasr — CHANGES_REQUESTED** [review]

File: cloudx/cmd_auth_test.go [hunk 7]
```diff
@@ -70,6 +70,17 @@ func TestAuthenticator(t *testing.T) {
...
 
+		t.Run("forced to reauthenticate on session expiration", func(t *testing.T) {
+			cmd := configAwareCmd(configDir)
+			expectSignInSuccess(t)
+			changeAccessToken(t, configDir)
+			var r bytes.Buffer
+			r.WriteString("n\n") // Your CLI session has expired. Do you wish to login again as <email>?
+			_, stderr, err := cmd.ExecDebug(t, &r, "list", "projects")
+			require.Error(t, err)
+			assert.Contains(t, stderr, "Your CLI session has expired. Do you wish to log in again as")
```
Comment (aeneasr) [inline cloudx/cmd_auth_test.go:81]: Can you please add the log in sequence here (see the other tests), to ensure that we are actually able to log in properly?

File: cloudx/handler.go [hunk 8]
```diff
@@ -231,6 +231,30 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
 
 	if len(c.SessionToken) > 0 {
+		client, err := newKratosClient()
+		if err != nil {
+			return nil, err
+		}
+		_, resp, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if resp.StatusCode == http.StatusUnauthorized || err != nil {
+			if h.isQuiet {
+				return nil, errors.New("can not sign in or sign up when flag --quiet is set")
```
Comment (aeneasr) [inline cloudx/handler.go:241]: Can you please add an error message that explicitly states that the session has expired? Please also add a test for this behaviour.

File: cloudx/handler.go [hunk 9]
```diff
@@ -231,6 +231,30 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
 	}
 
 	if len(c.SessionToken) > 0 {
+		client, err := newKratosClient()
+		if err != nil {
+			return nil, err
+		}
+		_, resp, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if resp.StatusCode == http.StatusUnauthorized || err != nil {
```
Comment (aeneasr) [inline cloudx/handler.go:239]: If `err != nil`, then this will panic because `resp` can be `nil`.

File: cloudx/handler.go [hunk 10]
```diff
@@ -231,6 +231,30 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
+			if h.isQuiet {
+				return nil, errors.New("can not sign in or sign up when flag --quiet is set")
+			}
+			ok, err := cmdx.AskScannerForConfirmation(fmt.Sprintf("Your CLI session has expired. Do you wish to log in again as \"%s\"?", c.IdentityTraits.Email), h.stdin, h.verboseErrWriter)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
```
Comment (aeneasr) [inline cloudx/handler.go:247]: What happens if `ok` is `false`? It looks like we now just ignore the error. Please fix this and add a test for this behavior!

File: cloudx/handler_test.go [hunk 11]
```diff
@@ -103,6 +103,16 @@ func configPasswordAwareCmd(configDir, password string) *cmdx.CommandExecuter {
...
 }
 
+func changeAccessToken(t *testing.T, configDir string) {
+	ac := readConfig(t, configDir)
+	ac.SessionToken = "12341234"
+	data, err := json.Marshal(ac)
+	require.NoError(t, err)
+	err = os.WriteFile(configDir, data, 0644)
+	require.NoError(t, err)
+
```
Comment (aeneasr) [inline cloudx/handler_test.go:113]:
```suggestion
	require.NoError(t, os.WriteFile(configDir, data, 0644))
```

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

File: cloudx/handler.go (hunk 10 above)
Comment (gen1us2k) [reply cloudx/handler.go:247]: I handled this case

File: cloudx/handler.go (hunk 9 above)
Comment (gen1us2k) [reply cloudx/handler.go]: I think that checking the session for the `nil` value would be better in that case.

File: cloudx/handler.go [hunk 12]
```diff
@@ -231,6 +231,31 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
 
 	if len(c.SessionToken) > 0 {
+		client, err := newKratosClient()
+		if err != nil {
+			return nil, err
+		}
+		sess, _, err := client.V0alpha2Api.ToSession(h.ctx).XSessionToken(c.SessionToken).Execute()
+		if sess == nil || err != nil {
+			if h.isQuiet {
+				return nil, errors.New("your session expired and you can not reauthenticate when flag --quiet is set")
```
Comment (icyphox) [inline cloudx/handler.go:241]:
```suggestion
				return nil, errors.New("Your session has expired and you cannot reauthenticate when the --quiet flag is set")
```

Comment (icyphox) [reply cloudx/handler.go:241]: Small nit. :)

File: cloudx/handler.go [hunk 13]
```diff
@@ -231,6 +231,31 @@ func (h *SnakeCharmer) EnsureContext() (*AuthContext, error) {
...
+				if err := h.SignOut(); err != nil {
+					return nil, err
+				}
+				c, err := h.Authenticate()
+				if err != nil {
+					return nil, err
+				}
+				return c, nil
+			}
+			return nil, errors.New("your session expired")
```
Comment (icyphox) [inline cloudx/handler.go:257]: Just for consistency:
```suggestion
			return nil, errors.New("Your session has expired")
```

**Review by aeneasr — APPROVED** [review]

**Pull request merged**